server/channels/store/sqlstore/session_store.go

```diff
@@ -154,7 +154,8 @@ func (me SqlSessionStore) GetSessionsWithActiveDeviceIds(userId string) ([]*mode
 			UserId = ? AND
 			ExpiresAt != 0 AND
 			? <= ExpiresAt AND
-			DeviceId != ''`
+			DeviceId != '' AND
+			DeviceId != COALESCE(Props->>'last_removed_device_id', '')`
```

Comment:
Not a blocker as I'm mot expecting this to happen a lot of times, but it would be interesting to see the impact of this change at the db level

Reply:
@agarciamontoro Is this something the current performance tests would catch? Is there anything special you would like to do with this?

Reply:
For specific queries like this, I'd say it's better to run manual tests, analyzing the output of `EXPLAIN (ANALYZE, BUFFERS)` on the old and new queries with a sizeable database. Performance tests would catch something like this if it introduces a bad regression (and if we have coverage for the feature in the load-test tool, of course)

Reply:
@agarciamontoro With 6600 sessions in the database running on my local machine
Before:
```
mattermost_test=# EXPLAIN (ANALYZE, BUFFERS) SELECT * FROM Sessions WHERE UserId = 'ngxjmp3cmffxixsa5q8fwe3bbh' AND ExpiresAt != 0 AND 1 <= ExpiresAt AND DeviceID != '';
                                                            QUERY PLAN                                                            
----------------------------------------------------------------------------------------------------------------------------------
 Bitmap Heap Scan on sessions  (cost=24.12..405.30 rows=335 width=377) (actual time=0.251..1.046 rows=351 loops=1)
   Recheck Cond: ((userid)::text = 'ngxjmp3cmffxixsa5q8fwe3bbh'::text)
   Filter: ((expiresat <> 0) AND (1 <= expiresat) AND ((deviceid)::text <> ''::text))
   Rows Removed by Filter: 149
   Heap Blocks: exact=38
   Buffers: shared hit=9 read=36
   ->  Bitmap Index Scan on idx_sessions_user_id  (cost=0.00..24.03 rows=500 width=0) (actual time=0.212..0.213 rows=500 loops=1)
         Index Cond: ((userid)::text = 'ngxjmp3cmffxixsa5q8fwe3bbh'::text)
         Buffers: shared hit=1 read=6
 Planning Time: 0.678 ms
 Execution Time: 1.149 ms
(11 rows)
```
After:
```
mattermost_test=# EXPLAIN (ANALYZE, BUFFERS) SELECT * FROM Sessions WHERE UserId = 'ngxjmp3cmffxixsa5q8fwe3bbh' AND ExpiresAt != 0 AND 1 <= ExpiresAt AND DeviceID != '' AND DeviceId != COALESCE(Props->>'last_removed_device_id', '');
                                                                                  QUERY PLAN                                                                                   
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Bitmap Heap Scan on sessions  (cost=24.12..407.80 rows=334 width=377) (actual time=0.168..0.839 rows=351 loops=1)
   Recheck Cond: ((userid)::text = 'ngxjmp3cmffxixsa5q8fwe3bbh'::text)
   Filter: ((expiresat <> 0) AND (1 <= expiresat) AND ((deviceid)::text <> ''::text) AND ((deviceid)::text <> COALESCE((props ->> 'last_removed_device_id'::text), ''::text)))
   Rows Removed by Filter: 149
   Heap Blocks: exact=38
   Buffers: shared hit=45
   ->  Bitmap Index Scan on idx_sessions_user_id  (cost=0.00..24.03 rows=500 width=0) (actual time=0.131..0.131 rows=500 loops=1)
         Index Cond: ((userid)::text = 'ngxjmp3cmffxixsa5q8fwe3bbh'::text)
         Buffers: shared hit=7
 Planning Time: 0.335 ms
 Execution Time: 0.948 ms
(11 rows)

```

Reply:
Looks good to me! It seems caching was a bit benefitial for the query run in the second place, but the plan is the exact same and we already have a filter in place, so I see no problem here. Thanks for taking the time to do this!